### ISSUE TITLE
There is no support for `weight_decay`/`momentum` in SGD for sparse tensors.

### ISSUE THREAD

**Issue by rrkarim** [pytorch/pytorch#30402]
## 🚀 Feature
I don't see why there is no support for `weight_decay` and `momentum` in SGD. There is a PR on this feature (#1305), but it has been never merged with no explanation. It looks like really easy patch to be implemented. Or maybe there it is not possible way to provide the exact same semantics with sparse tensors because of all the constraints on `coalesced` sparse tensors for some ops?

**Comment by SsnL** [comment]
IMO, generally the ambiguity is which entries should be decayed/momentum accelerated.
+ For dense tensor and sparse gradients:
  
  Changing all values in the tensor doesn't make much sense as this case is normally found in embedding training, where decaying / accelerating unused entries is rather weird (since the #iter between an entry gets used twice could be very big but we don't want that entry to, e.g., just decay to zero in between).

  Changing only the values in the gradients also feels wrong. This is different from the normal definition of decay/momentum, and it is unclear why only the values that are updated should be decayed.

+ For sparse tensor and sparse gradients:

  Changing all values in the tensor is not an option anymore (at least for momentum) as the base tensor should be sparse.

  Changing only the values in the gradients has the same concern as the above case.

**Comment by rrkarim** [comment]
@SsnL thanks for the response, I understand the concerns on the ambiguity and the lack of theory to justify the particular choice of update logic. But there is an implementation of `MomentumOptimizer` in tensorflow with the support for sparse tensors and, if I'm not wrong, in the docs they write that the method updates only the values in `grad` (your second point).

**Comment by rrkarim** [comment]
Though, unfortunately, I haven't found any analyses that could justify the effectiveness of this approach.

**Comment by rrkarim** [comment]
May be relatable: https://arxiv.org/pdf/1909.12778.pdf

**Comment by rrkarim** [comment]
Have implemented the version with the update of some values, maybe could be useful for someone:
```
def step(self, closure=None):
     """Performs a single optimization step.
 
     Arguments:
         closure (callable, optional): A closure that reevaluates the model
             and returns the loss.
     """
     loss = None
     if closure is not None:
         loss = closure()
 
     for group in self.param_groups:
         weight_decay = group['weight_decay']
         momentum = group['momentum']
         dampening = group['dampening']
         nesterov = group['nesterov']
 
         for p in group['params']:
             if p.grad is None:
                 continue
             d_p = p.grad.data
             if not d_p.is_sparse:
                 if weight_decay != 0:
                     d_p.add_(weight_decay, p.data)
                 if momentum != 0:
                     param_state = self.state[p]
                     if 'momentum_buffer' not in param_state:
                         buf = param_state['momentum_buffer'] = torch.clone(d_p).detach()
                     else:
                         buf = param_state['momentum_buffer']
                         buf.mul_(momentum).add_(1 - dampening, d_p)
                     if nesterov:
                         d_p = d_p.add(momentum, buf)
                     else:
                         d_p = buf 
 
                 p.data.add_(-group['lr'], d_p)
             else:
                 d_p = d_p.coalesce()
                 d_p_inds = d_p._indices()
                 d_p_values = d_p._values()
                 size = d_p.size()
 
                 def make_sparse(values):
                     constructor = d_p.new
                     if d_p_inds.dim() == 0 or values.dim() == 0:
                         return constructor().resize_as_(d_p)
                     return constructor(d_p_inds, values, size)
 
                 if weight_decay != 0:
                     param_values = p.data.sparse_mask(d_p)._values()
                     d_p_values.add_(param_values, weight_decay)
                 if momentum != 0:
                     param_state = self.state[p]
                     if "momentum_buffer" not in param_state:
                         buf = param_state["momemntum_state"] = torch.clone(d_p_values).detach()
                     else:
                         buf = param_state["momentum_buffer"]
                         buf.mul_(momentum).add_(1 - dampening, d_p_values)
                     if nesterov:
                         d_p_values = d_p_values.add(momentum, buf)
                     else:
                         d_p_values = buf 
 
                 p.data.add_(make_sparse(-group["lr"] * d_p_values))
 
         return loss

 ```

**Comment by karlhigley** [comment]
> Changing all values in the tensor doesn't make much sense as this case is normally found in embedding training, where decaying / accelerating unused entries is rather weird (since the #iter between an entry gets used twice could be very big but we don't want that entry to, e.g., just decay to zero in between).

is part of the answer to:

> it is unclear why only the values that are updated should be decayed.

Also: In this dense tensor/sparse gradient case, choosing an appropriate value of the `weight_decay` parameter currently depends on the size of the dataset, which very non-intuitive and makes training and tuning models unnecessarily difficult.

**Comment by karlhigley** [comment]
See also #1285 re: sparse optimizers and which parameters they should update
